src/core/model/table/dynamic_table_data.h

```diff
@@ -0,0 +1,82 @@
+#pragma once
+
+#include <cassert>
+#include <string>
+#include <unordered_set>
+#include <vector>
+
+#include "config/exceptions.h"
+#include "config/tabular_data/input_table_type.h"
+
+namespace model {
+
+struct DynamicTableData {
+private:
+    std::vector<std::vector<std::string>> columns_;
+    std::unordered_set<size_t> deleted_rows_{};
+
+public:
+    DynamicTableData(config::InputTable input_table) {
+        columns_.resize(input_table->GetNumberOfColumns());
+        while (input_table->HasNextRow()) {
+            std::vector<std::string> row = input_table->GetNextRow();
+            assert(row.size() == columns_.size());
+            for (size_t i = 0; i < row.size(); ++i) {
+                columns_[i].push_back(std::move(row[i]));
+            }
+        };
+    }
+
+    size_t GetNumRowsActual() const {
+        return columns_[0].size() - deleted_rows_.size();
+    }
+
+    size_t GetNumRowsTotal() const {
+        return columns_[0].size();
+    }
+
+    std::string const& GetValue(size_t row_index, size_t col_index) const {
+        assert(col_index < columns_.size() && row_index < columns_[col_index].size());
+        return columns_[col_index][row_index];
+    }
+
+    bool ValidateRowIndex(size_t idx) {
+        return !(deleted_rows_.contains(idx) || idx >= GetNumRowsTotal());
+    }
+
+    void Update(config::InputTable insert_data, config::InputTable update_data,
+                std::unordered_set<size_t> const& delete_data) {
+        for (size_t row_id : delete_data) {
+            for (size_t i = 0; i < columns_.size(); ++i) {
+                columns_[i][row_id] = {};
+            }
+            deleted_rows_.emplace(row_id);
+        }
+        if (insert_data != nullptr) {
+            while (insert_data->HasNextRow()) {
+                std::vector<std::string> row = insert_data->GetNextRow();
+                assert(row.size() == columns_.size());
+                for (size_t i = 0; i < row.size(); ++i) {
+                    columns_[i].push_back(std::move(row[i]));
+                }
+            }
+        }
+        if (update_data != nullptr) {
+            while (update_data->HasNextRow()) {
+                std::vector<std::string> row = update_data->GetNextRow();
+                size_t row_id = row.empty() ? 0 : std::stoull(row.front());
+                if (deleted_rows_.contains(row_id)) {
+                    throw config::ConfigurationError(
+                            "Attempt to update a deleted row during processing of update "
+                            "operations");
+                }
+                assert(row.size() == columns_.size() + 1);
```

Comment:
`assert` is only used for debugging; this is data from the user, so it should be checked properly.